cmd/gateway-azure.go

```diff
@@ -307,96 +307,102 @@ func (a *azureObjects) DeleteBucket(bucket string) error {
+		if resp.NextMarker == "" {
 			break
 		}
-	}
-	result.Prefixes = resp.BlobPrefixes
-	return result, nil
-}
 
-// ListObjectsV2 - list all blobs in Azure bucket filtered by prefix
-func (a *azureObjects) ListObjectsV2(bucket, prefix, continuationToken string, fetchOwner bool, delimiter string, maxKeys int) (result ListObjectsV2Info, err error) {
-	resp, err := a.client.ListBlobs(bucket, storage.ListBlobsParameters{
-		Prefix:     prefix,
-		Marker:     continuationToken,
-		Delimiter:  delimiter,
-		MaxResults: uint(maxKeys),
-	})
-	if err != nil {
-		return result, azureToObjectError(traceError(err), bucket, prefix)
-	}
-	// If NextMarker is not empty, this means response is truncated and NextContinuationToken should be set
-	if resp.NextMarker != "" {
-		result.IsTruncated = true
-		result.NextContinuationToken = resp.NextMarker
+		marker = resp.NextMarker
 	}
-	for _, object := range resp.Blobs {
-		if strings.HasPrefix(object.Name, globalMinioSysTmp) {
-			continue
-		}
-		t, e := time.Parse(time.RFC1123, object.Properties.LastModified)
-		if e != nil {
-			continue
+
+	if version == 2 {
+		var resultV2 ListObjectsV2Info
+		resultV2.Objects = objects
+		resultV2.Prefixes = prefixes
+		if marker != "" {
+			resultV2.IsTruncated = true
+			resultV2.NextContinuationToken = marker
+		} else {
+			resultV2.ContinuationToken = markerContinuationToken
 		}
-		result.Objects = append(result.Objects, ObjectInfo{
-			Bucket:          bucket,
-			Name:            object.Name,
-			ModTime:         t,
-			Size:            object.Properties.ContentLength,
-			ETag:            azureToS3ETag(object.Properties.Etag),
-			ContentType:     object.Properties.ContentType,
-			ContentEncoding: object.Properties.ContentEncoding,
-		})
-	}
 
-	// Remove minio.sys.tmp prefix.
-	for i, prefix := range resp.BlobPrefixes {
-		if prefix == globalMinioSysTmp {
-			resp.BlobPrefixes = append(resp.BlobPrefixes[:i], resp.BlobPrefixes[i+1:]...)
-			break
+		result = resultV2
+	} else {
+		var resultV1 ListObjectsInfo
+		resultV1.Objects = objects
+		resultV1.Prefixes = prefixes
+		if marker != "" {
+			resultV1.IsTruncated = true
+			resultV1.NextMarker = marker
 		}
+
+		result = resultV1
 	}
-	result.Prefixes = resp.BlobPrefixes
+
 	return result, nil
 }
 
+// ListObjects - lists all blobs on azure with in a container filtered by prefix
+// and marker, uses Azure equivalent ListBlobs.
+func (a *azureObjects) ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	var r interface{}
+	r, err = a.listObjects(bucket, prefix, marker, delimiter, maxKeys, 1)
+	result = r.(ListObjectsInfo)
+
+	return result, err
+}
+
+// ListObjectsV2 - list all blobs in Azure bucket filtered by prefix
+func (a *azureObjects) ListObjectsV2(bucket, prefix, continuationToken string, fetchOwner bool, delimiter string, maxKeys int) (result ListObjectsV2Info, err error) {
```

Comment:
Simpler implementation (i.e you can avoid `a.listObjects`, passing of version number and returning interface) by doing this
```
func (a *azureObjects) ListObjectsV2() {
    resultV1 := a.ListObjects() // Call V1
    resultV2 := listObjectsV1ToV2(resultV1)
    return resultV1
}
```

Reply:
Unable to understand why it is simpler than avoiding interface.